cmake/modules/FindStdFilesystem.cmake

```diff
@@ -0,0 +1,37 @@
+set(_std_filesystem_test_src
+  ${CMAKE_CURRENT_LIST_DIR}/FindStdFilesystem_test.cc)
+
+macro(try_std_filesystem_library _library _result)
+  try_compile(_std_filesystem_compiles
+    ${CMAKE_CURRENT_BINARY_DIR}
+    SOURCES ${_std_filesystem_test_src}
+    CMAKE_FLAGS -DCMAKE_CXX_FLAGS="-std=c++17"
+    LINK_LIBRARIES ${_library})
+  if(_std_filesystem_compiles)
+    set(${_result} ${_library})
+  endif()
+endmacro()
+
+
+if(NOT StdFilesystem_LIBRARY)
+  try_std_filesystem_library("stdc++fs" StdFilesystem_LIBRARY)
+endif()
+if(NOT StdFilesystem_LIBRARY)
+  try_std_filesystem_library("c++experimental" StdFilesystem_LIBRARY)
+endif()
+if(NOT StdFilesystem_LIBRARY)
+  try_std_filesystem_library("c++fs" StdFilesystem_LIBRARY)
```

Comment:
@wjwithagen i think this should address the issue you mentioned before.

Reply:
Doesn't cleanly apply to mimic, as was to be expected.
So I'll put this on the TODO, and see what is essential to backport to get `c++experimental` fixed.

Reply:
i will try to prepare a backport once this PR gets merged.

Reply:
@tchaikov 
Oke,
Question is: how much do you want to backport? Just code to fix the problem, or the whole batch?

Reply:
the former.

Reply:
@tchaikov 
This is the linker error:
```
/usr/bin/c++  -O2 -pipe  -fstack-protector -DLDAP_DEPRECATED -fno-strict-aliasing -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -Wno-unknown-pragmas -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -O2 -pipe -fstack-protector -DLDAP_DEPRECATED -fno-strict-aliasing -ftemplate-depth-1024 -Wnon-virtual-dtor -Wno-unknown-pragmas -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -fdiagnostics-color=auto -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -O2 -pipe -fstack-protector -DLDAP_DEPRECATED -fno-strict-aliasing  -Wl,--export-dynamic CMakeFiles/ceph_test_admin_socket_output.dir/test_admin_socket_output.cc.o CMakeFiles/ceph_test_admin_socket_output.dir/admin_socket_output.cc.o CMakeFiles/ceph_test_admin_socket_output.dir/admin_socket_output_tests.cc.o  -o ../../bin/ceph_test_admin_socket_output  -L/usr/local/lib -Wl,-rpath,/usr/local/lib:/usr/ports/net/ceph13/work/.build/lib:/usr/local/lib/nss: ../../lib/libceph-common.so.0 -lc++experimental ../../lib/libjson_spirit.a ../../lib/libcommon_utf8.a ../../lib/liberasure_code.a /usr/local/lib/libboost_thread.so /usr/local/lib/libboost_chrono.so /usr/local/lib/libboost_atomic.so /usr/local/lib/libboost_random.so /usr/local/lib/libboost_system.so /usr/local/lib/libboost_program_options.so /usr/local/lib/libboost_date_time.so /usr/local/lib/libboost_iostreams.so /usr/local/lib/libboost_regex.so -lexecinfo /usr/local/lib/nss/libssl3.so /usr/local/lib/nss/libsmime3.so /usr/local/lib/nss/libnss3.so /usr/local/lib/nss/libnssutil3.so /usr/local/lib/libplds4.so /usr/local/lib/libplc4.so /usr/local/lib/libnspr4.so -lssl -lcrypto -pthread
/usr/bin/ld: cannot find -lc++experimental
c++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[3]: *** [src/test/CMakeFiles/ceph_test_admin_socket_output.dir/build.make:140: bin/ceph_test_admin_socket_output] Error 1
```
My solution is to exclude building this testprogram during the package building.

Reply:
@tchaikov 
:
This is the real patch I'm using, (previous was bogus)
```
--- src/test/CMakeLists.txt
+++ src/test/CMakeLists.txt
@@ -515,6 +515,8 @@ if(HAVE_BLKID)
     )
 endif(HAVE_BLKID)

+if(NOT FREEBSD OR EXISTS "/usr/lib/libc++experimental.a")
+
 # ceph_test_admin_socket_output

 add_executable(ceph_test_admin_socket_output
@@ -534,6 +536,7 @@ endif()
 install(TARGETS
   ceph_test_admin_socket_output
   DESTINATION ${CMAKE_INSTALL_BINDIR})
+endif()

 #make check starts here
```

Reply:
i think replacing `libc++experimental.a` with `libc++fs` should fix your issue. and i am not sure what the context of your error message or your patch is in your comment? again, probably you could offer the clang version, what Ceph version you are building, if you are using libstdc++  or libc++ ...

i can hardly help if you just post the error message here.

Reply:
@tchaikov 
Sorry I was unclear:
Mimic 13.2.1
FreeBSD 11.2 (Which is the most recent stable version)
Clang version 6.0.0 (tags/RELEASE_600/final 326565) (based on LLVM 6.0.0)
Using the standard libc that Clang uses, which is `/usr/lib/libc++`
`libc++fs` is not available, perhaps I could use a package.

I chose to not include the test, hence my patch, as the test is the only program that suffers from this missing lib. So for packaging it is not required, and excluding from running tests is not nice, but if needed so be it. 

Note that on 12.x libc++experimental will be available per default, and the problem goes away by itself.

Reply:
ahh, i see the full picture now!

as the clang doc explains, the libc++experimental is optional. and libc++fs is only available after clang-7. and FreeBSD 11.2 does not have it, see https://reviews.freebsd.org/D10840 .

so this change will address the issue *after* FreeBSD moves to clang-7.

for now, we should disable the test if `std::filesystem` or `std::experimental::filesystem` is not available. will update the patch to do so.

Reply:
Yup, that is the reason why it is already there on 12.0-CURRENT, so that'll be oke. 
I don't see any backport notion, and given that the commit was over  a year ago, I would not expect it to land in 11.x. until clang-7 like you say.
So:
`if(NOT FREEBSD OR EXISTS "/usr/lib/libc++experimental.a")`
does the job for me.